text/0099-transaction-extension-version.md

```diff
@@ -0,0 +1,81 @@
+# RFC-0099: Introduce a transaction extension version
+
+|                 |                                                                                             |
+| --------------- | ------------------------------------------------------------------------------------------- |
+| **Start Date**  | 03 July 2024                                                                               |
+| **Description** | Introduce a versioning for transaction extensions.                               |
+| **Authors**     | Bastian Köcher                                                                              |
+
+## Summary
+
+This RFC proposes a change to the extrinsic format to include a transaction extension version.
+
+## Motivation
+
+The extrinsic format supports to be extended with transaction extensions. These transaction extensions are runtime specific and can be different per chain. Each transaction extension can add data to the extrinsic itself or extend the signed payload.
+This means that adding a transaction extension is breaking the chain specific extrinsic format. A recent example was the introduction of the [`CheckMetadatHash`](https://github.com/polkadot-fellows/runtimes/pull/337) to Polkadot and all its system chains.
+As the extension was adding one byte to the extrinsic, it broke a lot of tooling. By introducing an extra version for the transaction extensions it will be possible to introduce changes to these transaction extensions while still being backwards compatible. 
+Based on the version of the transaction extensions, each chain runtime could decode the extrinsic correctly and also create the correct signed payload.
+
+## Stakeholders
+
+- Runtime users
+- Runtime devs
+- Wallet devs
+
+## Explanation
+
+[RFC84](https://github.com/paritytech/polkadot-sdk/issues/2415) introduced the extrinsic format `5`. The idea is to piggyback onto this change of the extrinsic format to add the extra version for the transaction extensions. If required, this could also come 
+as extrinsic format `6`, but `5` is not yet deployed anywhere. 
+
+The extrinsic format supports the following types of transactions:
+- `Bare`: Does not add anything to the extrinsic.
+- `Signed`: `(Address, Signature, Extensions)`
+- `General`: `Extensions`
+
+The `Signed` and `General` transaction would change to:
+
+- `Signed`: `(Address, Signature, Version, Extensions)`
+- `General`: `(Version, Extensions)`
+
+The `Version` being a SCALE encoded compact `u32` representing the version of the transaction extensions.
```

Comment:
should we just use u8?
cons:
- only up to 256 versions. but it is very unlikely a runtime needs to support that many versions and we can recycle old versions if necessary
pros:
- compact encoding using 2 bits to indicate encoding type and that means for number >=64, it will require 2 bytes to encode instead of 1. it is likely some chains will have that many versions and have 1 byte additional overhead

Reply:
I thought about this as well. I also know the limitation of only having `64` versions until we need two bytes :D 

I mean we can go with `u8` and just bump the tx main version again if we need more :see_no_evil: